packages/react-sdk/src/components/BackgroundFilters/BackgroundFilters.tsx

```diff
@@ -194,167 +213,145 @@ export const BackgroundFiltersProvider = (
+    {
+      width: 1920,
+      height: 1080,
+    },
+  );
 
-    const filter = canvasRef.captureStream();
-    resolveFilter(filter);
-    return () => {
-      disposeOfMediaStream(filter);
-    };
-  }, [canvasRef]);
+  const start = useCallback(
+    (ms: MediaStream, onError?: (error: any) => void) => {
+      let outputStream: MediaStream | undefined;
+      let renderer: Renderer | undefined;
+
+      const output = new Promise<MediaStream>((resolve, reject) => {
+        if (!backgroundFilter) {
+          reject(new Error('No filter specified'));
+          return;
+        }
+
+        const videoEl = videoRef.current;
+        const canvasEl = canvasRef.current;
+        const bgImageEl = bgImageRef.current;
+
+        if (!videoEl || !canvasEl || (backgroundImage && !bgImageEl)) {
+          // You should start renderer in effect or event handlers
+          reject(new Error('Renderer started before elements are ready'));
+          return;
+        }
+
+        videoEl.srcObject = ms;
+        videoEl.play().then(() => {
+          const [track] = ms.getVideoTracks();
+
+          if (!track) {
+            reject(new Error('No video tracks in input media stream'));
+          }
+
+          const trackSettings = track.getSettings();
+          flushSync(() =>
+            setVideoSize({
+              width: trackSettings.width ?? 0,
+              height: trackSettings.height ?? 0,
+            }),
+          );
+          renderer = createRenderer(
+            tfLite,
+            videoEl,
+            canvasEl,
+            {
+              backgroundFilter,
+              backgroundBlurLevel,
+              backgroundImage: bgImageEl ?? undefined,
+            },
+            onError,
+          );
+          outputStream = canvasEl.captureStream();
+          resolve(outputStream);
+        });
+      });
+
+      return {
+        output,
+        stop: () => {
+          renderer?.dispose();
+          videoRef.current && (videoRef.current.srcObject = null);
+          outputStream && disposeOfMediaStream(outputStream);
+        },
+      };
+    },
+    [backgroundBlurLevel, backgroundFilter, backgroundImage, tfLite],
+  );
 
-  return (
-    <div
-      className="str-video__background-filters"
-      style={{
-        width: `${width}px`,
-        height: `${height}px`,
-      }}
-    >
-      {mediaStream && isPlaying && (
-        <RenderPipeline
-          tfLite={tfLite}
-          videoRef={videoRef}
-          canvasRef={canvasRef}
-          backgroundImageRef={bgImageRef}
-        />
-      )}
+  const children = (
+    <div className="str-video__background-filters">
       <video
         className={clsx(
           'str-video__background-filters__video',
-          height > width && 'str-video__background-filters__video--tall',
+          videoSize.height > videoSize.width &&
+            'str-video__background-filters__video--tall',
         )}
-        ref={setVideoRef}
-        autoPlay
```

Comment:
Did we intentionally remove `autoPlay` because we want to control this via `ref.play()`?

Reply:
Yes, I just didn't think it's necessary since we always call `ref.play()`. Are there any advantages to having `autoPlay`?

Reply:
no, I think it is good like this. `autoPlay` should immediately kick-start the playback once `srcObject` is set and there is enough metadata loaded.